src/query/RTree.hpp

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <vector>
+#include "mesh/BoundingBox.hpp"
+#include "mesh/Edge.hpp"
+#include "mesh/SharedPointer.hpp"
+#include "mesh/Triangle.hpp"
+#include "mesh/Vertex.hpp"
+
+namespace precice {
+namespace query {
+
+class MeshIndices;
+
+struct MatchType {
+  double distance;
+  int    index;
+
+  MatchType() = default;
+  MatchType(double d, int i)
+      : distance(d), index(i){};
+  constexpr bool operator<(MatchType const &other) const
+  {
+    return distance < other.distance;
+  };
+};
+
+class rtree {
+public:
+  /// Get the closest vertex/edge/triangle to a vertex, return indices and distance
+  static std::vector<MatchType> getClosestVertex(const mesh::Vertex &source, const mesh::PtrMesh &targetMesh, int n = 1);
+  static std::vector<MatchType> getClosestEdge(const mesh::Vertex &source, const mesh::PtrMesh &targetMesh, int n = 1);
+  static std::vector<MatchType> getClosestTriangle(const mesh::Vertex &source, const mesh::PtrMesh &targetMesh, int n = 1);
```

Comment:
We could use 3 different MatchTypes here.
A simple solution could be to tag the different types:

```cpp
template<typename Tag>
struct MatchType { /*  ... */ };

using VertexMatch = MatchType<struct VertexMatchTag>;
using EdgeMatch = MatchType<struct EdgeMatchTag>;
//...
static std::vector<VertexMatch> getClosestVertex(const mesh::Vertex &source, const mesh::PtrMesh &targetMesh, int n = 1);
```